meeko/molsetup.py

```diff
@@ -4,623 +4,451 @@
-    
-
-# TODO modify so that there are no more dictionaries and only list/arrays (fix me)
-# methods like add_x,del_x,get_x will deal with indexing (fix me)
-# the goal is to remove dictionaries (fix me)
+# region DEFAULT VALUES
+DEFAULT_PDBINFO = None
+DEFAULT_CHARGE = 0.0
+DEFAULT_ELEMENT = None
+DEFAULT_ATOM_TYPE = None
+DEFAULT_IS_IGNORE = False
+DEFAULT_IS_CHIRAL = False
 
-# (fix me)) provide infrastructure to calculate "differential" modifications,
-# like tautomers: the class calculating tautomers will copy and modify the setup
-# to swap protons and change the atom types (and possibly rotations?), then
-# there will e a function to extract only differences (i.e., atom types and coordinates,
-# in the case of tautomers) and store them to make it a switch-state
+DEFAULT_BOND_ORDER = 0
+DEFAULT_BOND_ROTATABLE = False
 
-# TODO use only 1D arrays (C)
+DEFAULT_RING_CORNER_FLIP = False
+DEFAULT_RING_GRAPH = defaultdict
+DEFAULT_RING_IS_AROMATIC = False
+# endregion
 
 
-# TODO update add_atom to not to specify neighbors, which should be defined only when a bond is created?
-# TODO change all attributes_to_copy to have underscore ?
-
 class MoleculeSetup:
-    """ mol: molecule structurally prepared with explicit hydrogens
 
-        the setup provides:
-            - data storage
-            - SMARTS matcher for all atom typers
-    """
+    # region CLASS CONSTANTS
+    PSEUDOATOM_ELEMENT = 0
+    # endregion
 
-    def __init__(self):
+    def __init__(self, name: str = None, is_sidechain: bool = False):
+        # Molecule Setup Identity
+        self.name = name
+        self.is_sidechain = is_sidechain
+        self.true_atom_count = 0
+        self.pseudo_atom_count = 0
 
-        self.atom_pseudo = []
-        self.coord = OrderedDict()  # FIXME all OrderedDict should be converted to lists?
-        self.charge = OrderedDict()
-        self.pdbinfo = OrderedDict()
-        self.atom_type = OrderedDict()
-        self.atom_params = {}
-        self.dihedral_interactions = list()
-        self.dihedral_partaking_atoms = dict()
-        self.dihedral_labels = dict()
-        self.atom_ignore = OrderedDict()
-        self.chiral = OrderedDict()
-        self.atom_true_count = 0
-        self.graph = OrderedDict()
-        self.bond = OrderedDict()
-        self.element = OrderedDict()
-        self.interaction_vector = OrderedDict()
-        self.flexibility_model = {}
-        self.ring_closure_info = {
-            "bonds_removed": [],
-            "pseudos_by_atom": {},
-        }
-        self.restraints = []
-        self.is_sidechain = False
-        self.rmsd_symmetry_indices = ()
+        # Tracking atoms and bonds
+        self.atoms: list[Atom] = []
+        self.bond_info = {}
         self.rings = {}
-        self.rings_aromatic = []
-        self.atom_to_ring_id = defaultdict(list)
-        self.ring_corners = {}  # used to store corner flexibility
-        self.name = None
         self.rotamers = []
```

Comment:
Is this where we really want rotamers? Does this look like an appropriate level, or should they be specific to RDKit or something like that

Reply:
I think this is the right place.
Rotamers define values (angles) that are associated with any number of bonds. For example, we can have a molecule with 3 rotatable bonds, and have any number of rotamers. Each rotamer can define angles for any number of bonds. For example:
rotamer_1: bond_1=1.0, bond_2=1.5
rotamer_2: bond_1=3.0